Resolve variant::toJson() recursive call with DecimalType

Krishna-Prasad-P-V opened the pull request:
Resolves #5338

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -257,19 +259,22 @@ std::string variant::toJson(const TypePtr& type) const {
       return target;
     }
-    case TypeKind::HUGEINT:
-      VELOX_CHECK(type && type->isLongDecimal());
```
This should remain a `VELOX_CHECK` since int128_t native type is not officially supported.

Krishna-Prasad-P-V replied:
Resolved.

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -221,11 +221,12 @@ std::string variant::toJson(const TypePtr& type) const {
           b += ",";
         }
-        b += v.toJson();
+        b += v.toJson(type->childAt(idx++));
```
We need to verify that the typeKind of `type->childAt(idx++)` matches typeKind of `v`

Krishna-Prasad-P-V replied:
Added the check. `VELOX_USER_CHECK_EQ(
            v.kind(), type->childAt(idx)->kind(), "Not same TypeKind");`

majetideepak commented on velox/type/tests/VariantTest.cpp:
```diff
@@ -272,3 +272,22 @@ TEST(VariantTest, opaqueSerializationNotRegistered) {
   EXPECT_THROW(opaqueAfterRegistration.serialize(), VeloxException);
 }
+
+TEST(VariantTest, toJson) {
```
Add two negative test cases.
```
auto rowTypePtr = ROW({{"c0", DECIMAL(10, 3)}});
EXPECT_THROW(
      variant::row({variant((int128_t)123456789)}).toJson(rowTypePtr), "Not same TypeKind");
rowTypePtr = ROW({{"c0", DECIMAL(20, 3)}});
EXPECT_THROW(
      variant::row({variant((int64_t)123456789)}).toJson(rowTypePtr), "Not same TypeKind");
```

Krishna-Prasad-P-V replied:
Added the provided negative tests.
```
  EXPECT_THROW(
      variant::row({variant((int128_t)123456789)}).toJson(rowTypePtr),
      VeloxUserError);
  rowTypePtr = ROW({{"c0", DECIMAL(20, 3)}});
  EXPECT_THROW(
      variant::row({variant((int64_t)123456789)}).toJson(rowTypePtr),
      VeloxUserError);
```

Krishna-Prasad-P-V replied:
Tests ran fine with the new negative cases as expected.
```
Running main() from /root/Presto/velox/_build/debug/_deps/gtest-src/googletest/src/gtest_main.cc
Note: Google Test filter = VariantTest.toJson
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from VariantTest
[ RUN      ] VariantTest.toJson
WARNING: Logging before InitGoogleLogging() is written to STDERR
E0626 08:52:28.771492  8439 Exceptions.h:68] Line: /root/Presto/velox/velox/type/Variant.cpp:229, Function:toJson, Expression: v.kind() == type->childAt(idx)->kind() (HUGEINT vs. BIGINT) Not same TypeKind, Source: USER, ErrorCode: INVALID_ARGUMENT
E0626 08:52:28.780982  8439 Exceptions.h:68] Line: /root/Presto/velox/velox/type/Variant.cpp:229, Function:toJson, Expression: v.kind() == type->childAt(idx)->kind() (BIGINT vs. HUGEINT) Not same TypeKind, Source: USER, ErrorCode: INVALID_ARGUMENT
[       OK ] VariantTest.toJson (9 ms)
[----------] 1 test from VariantTest (9 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (10 ms total)
[  PASSED  ] 1 test.
```

Krishna-Prasad-P-V commented:
@majetideepak - Kindly review this pull request as I have addressed the review comments.

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -236,11 +239,14 @@ std::string variant::toJson(const TypePtr& type) const {
           b += ",";
         }
-        b += v.toJson();
+        VELOX_USER_CHECK_EQ(
```
This check can be in the `toJson(const TypePtr& type)` method.
We can probably make the error message more descriptive.
```
VELOX_USER_CHECK_EQ(
            v.kind(), type->kind(), "variant TypeKind {} does not match input TypeKind {}", v.kind(), type->kind());
```

Krishna-Prasad-P-V replied:
We create `v` within the foreach loop for ROW and ARRAY TypeKinds. So we can't keep this check at toJson(). `v` would be inaccessible.

majetideepak replied:
It would be `this.kind()` instead of `v`.

Krishna-Prasad-P-V replied:
Changes the error message, but the check remains at same place.

majetideepak replied:
See my comment above.

Krishna-Prasad-P-V replied:
Fixed as above.

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -221,11 +221,18 @@ std::string variant::toJson(const TypePtr& type) const {
       std::string b{};
       b += "[";
       bool first = true;
+      uint32_t idx = 0;
```
We should check the number of children match as well along with a test.

Krishna-Prasad-P-V replied:
Added the checks in toJson() function.

Krishna-Prasad-P-V commented:
@majetideepak - Kindly review the PR. Addressed the comments.

FYI, local test run outputs
```
root@batch1:~/Presto/velox# /root/Presto/velox/_build/debug/velox/type/tests/velox_type_test --gtest_filter=VariantTest.toJson
Running main() from /root/Presto/velox/_build/debug/_deps/gtest-src/googletest/src/gtest_main.cc
Note: Google Test filter = VariantTest.toJson
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from VariantTest
[ RUN      ] VariantTest.toJson
WARNING: Logging before InitGoogleLogging() is written to STDERR
E0627 06:14:15.072263 232899 Exceptions.h:68] Line: /root/Presto/velox/velox/type/Variant.cpp:207, Function:toJson, Expression: this->kind() == type->kind() (HUGEINT vs. BIGINT) variant TypeKind HUGEINT does not match input TypeKind BIGINT, Source: USER, ErrorCode: INVALID_ARGUMENT
E0627 06:14:15.072731 232899 Exceptions.h:68] Line: /root/Presto/velox/velox/type/Variant.cpp:207, Function:toJson, Expression: this->kind() == type->kind() (BIGINT vs. HUGEINT) variant TypeKind BIGINT does not match input TypeKind HUGEINT, Source: USER, ErrorCode: INVALID_ARGUMENT
E0627 06:14:15.072976 232899 Exceptions.h:68] Line: /root/Presto/velox/velox/type/Variant.cpp:214, Function:toJson, Expression: totalIdxForVariantChild == totalIdxForInputChild (3 vs. 1) variant children count 3 does not match input children count 1, Source: USER, ErrorCode: INVALID_ARGUMENT
E0627 06:14:15.073258 232899 Exceptions.h:68] Line: /root/Presto/velox/velox/type/Variant.cpp:214, Function:toJson, Expression: totalIdxForVariantChild == totalIdxForInputChild (2 vs. 3) variant children count 2 does not match input children count 3, Source: USER, ErrorCode: INVALID_ARGUMENT
[       OK ] VariantTest.toJson (1 ms)
[----------] 1 test from VariantTest (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1 ms total)
[  PASSED  ] 1 test.
```

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -192,10 +192,32 @@ std::string variant::toJson(const TypePtr& type) const {
   // and other basic operations. Stringification logic should not be specific
   // to variants; it should be consistent for all map representations
 
+  static uint32_t totalIdxForVariantChild;
```
We don't need these two variables.

Krishna-Prasad-P-V replied:
Removed.

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -192,10 +192,32 @@ std::string variant::toJson(const TypePtr& type) const {
+      this->kind(),
+      type->kind());
+
+  VELOX_USER_CHECK_EQ(
```
these checks must go into ROW and Array toJson() function.

Krishna-Prasad-P-V replied:
Changed it to be used with ROW and ARRAY cases in toJson() function.

Krishna-Prasad-P-V commented:
@majetideepak - We have test failures with the check we did for kind_ in variant::toJson() function. Looks like there are cases where they call toJson() from Expressions.h and this is causing the issue. Should I revert this check to ROW and ARRAY cases as before? With those changes, tests ran fine and good. Let me know.

Issues seen:
```
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] TypedExprSerDeTest.constant
[  FAILED  ] TypedExprSerDeTest.lambda
```

New code where the issue is seen:
```
  VELOX_USER_CHECK_EQ(
      this->kind(),
      type->kind(),
      "variant TypeKind {} does not match input TypeKind {}",
      this->kind(),
      type->kind());

// this->kind() is received as facebook::velox::TypeKind::INTEGER
// type->kind() is receives as facebook::velox::TypeKind::BIGINT
```

Krishna-Prasad-P-V commented on velox/type/Variant.cpp:
```diff
@@ -221,11 +228,20 @@ std::string variant::toJson(const TypePtr& type) const {
+      uint32_t totalIdxForVariantChild = row.size();
+      uint32_t totalIdxForInputChild = type->size();
+      VELOX_USER_CHECK_EQ(
+          totalIdxForVariantChild,
```
Fixed.

majetideepak replied:
Did not push this change?

Krishna-Prasad-P-V replied:
Yes. Pushed the changes now.

majetideepak commented:
> TypedExprSerDeTest.constant

The tests seem to have an issue. For the above test can you change `127 ->  static_cast<int64_t>(127)`?
The others test need `10 -> static_cast<int64_t>(10)`

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -236,11 +250,18 @@ std::string variant::toJson(const TypePtr& type) const {
       std::string b{};
       b += "[";
       bool first = true;
+      uint32_t idx = 0;
```
Array can have only a single elementType as its child.
We should enforce that here and remove the `for` loop below.

Krishna-Prasad-P-V replied:
@majetideepak - I did the change to enforce it takes only single element type into consideration for ARRAY, but the for loop remains. Could you please confirm if this is sufficient here?
```
    case TypeKind::ARRAY: {
      auto& array = value<TypeKind::ARRAY>();
      std::string b{};
      b += "[";
      bool first = true;
      uint32_t idx = 0;
      auto arrayElementType = type->childAt(0);
      for (uint32_t idx = 0; idx < array.size(); ++idx)
      {
        if(!first) {
          b += ",";
        }
        b += array.at(idx).toJson(arrayElementType);
        first = false;
      }
      b += "]";
      return b;
    }
```

I also did the testing and works fine
```
  // Array variant tests
  auto arrayTypePtr = ARRAY(DECIMAL(20, 3));
  EXPECT_EQ(
      "[123456.789,634565.464,234545.278]",
      variant::array({variant((int128_t)123456789),
                      variant((int128_t)634565464),
                      variant((int128_t)234545278)})
          .toJson(arrayTypePtr));
  EXPECT_THROW( 
      variant::array({variant((int128_t)123456789),
                      variant((std::string) "Array should not support multiple element types.")})
          .toJson(arrayTypePtr),
      VeloxUserError);
```

majetideepak replied:
Why do we need the for loop for Array type? Can you explain?

Krishna-Prasad-P-V replied:
For example, suppose we have 3 elements of a type kind, then the toJson() function should run for each element. This is what my understanding is, isn't it?
This is example I tested.
```
auto arrayTypePtr = ARRAY(DECIMAL(20, 3));
  EXPECT_EQ(
      "[123456.789,634565.464,234545.278]",
      variant::array({variant((int128_t)123456789),
                      variant((int128_t)634565464),
                      variant((int128_t)234545278)})
          .toJson(arrayTypePtr));
```

majetideepak replied:
@Krishna-Prasad-P-V you are right. Thanks for the example.

yzhang1991 commented:
Hi @Krishna-Prasad-P-V, is this ready for another review?

majetideepak commented:
@Krishna-Prasad-P-V I see some tests failing as well. Likely related to this PR. Can you take a look?

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -227,11 +241,18 @@ std::string variant::toJson(const TypePtr& type) const {
+      VELOX_USER_CHECK_EQ(
+          array.size(),
+          type->size(),
+          "variant children count {} does not match input children count {}",
```
`input` -> `input type`

Krishna-Prasad-P-V commented on velox/type/Variant.cpp:
```diff
@@ -212,11 +219,18 @@ std::string variant::toJson(const TypePtr& type) const {
+      VELOX_USER_CHECK_EQ(
+          row.size(),
+          type->size(),
+          "variant children count {} does not match input children count {}",
```
Fixed.

Krishna-Prasad-P-V commented:
> @Krishna-Prasad-P-V I see some tests failing as well. Likely related to this PR. Can you take a look?

Yes, I have all those test failures with the latest push.

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -183,6 +183,15 @@ std::string variant::toJson(const TypePtr& type) const {
     return "null";
   }
 
+  if (type != nullptr) {
```
This check should be `VELOX_USER_CHECK(type != nullptr)` and at the top.

majetideepak replied:
I see why you need this check. `type` can be a `nullptr`.
It is not ideal to allow type to be a nullptr. It makes the implementation clunky and bug prone.
Example: `type->isDate()` check below is missing the nullptr check.
Let me look into this further.

majetideepak commented on velox/type/Variant.cpp:
```diff
@@ -227,11 +243,13 @@ std::string variant::toJson(const TypePtr& type) const {
       b += "[";
       bool first = true;
+      uint32_t idx = 0;
+      auto arrayElementType = (type != nullptr) ? type->childAt(0) : nullptr;
```
This can simply be `auto arrayElementType = type->childAt(0)` since we have the null check on top.

majetideepak commented on velox/type/tests/VariantTest.cpp:
```diff
@@ -285,3 +286,66 @@ TEST(VariantTest, opaqueSerializationNotRegistered) {
+      ROW({{"c0", DECIMAL(20, 1)}, {"c1", BOOLEAN()}, {"c3", VARCHAR()}});
+  EXPECT_EQ(
+      "[1234567890.1,true,\"test works fine\"]",
+      variant::row({variant((int128_t)12345678901),
```
static_cast<int128_t> here and below.

Krishna-Prasad-P-V replied:
Fixed.

majetideepak commented:
@Krishna-Prasad-P-V we decided here https://github.com/facebookincubator/velox/discussions/4069#discussioncomment-6488542 to remove `toJson()` signature and make `toJson(const TypePtr& type)` as the only supported signature. Can you use this PR as a reference https://github.com/facebookincubator/velox/pull/5720 and implement this change?

Krishna-Prasad-P-V commented:
> @Krishna-Prasad-P-V we decided here [#4069 (comment)](https://github.com/facebookincubator/velox/discussions/4069#discussioncomment-6488542) to remove `toJson()` signature and make `toJson(const TypePtr& type)` as the only supported signature. Can you use this PR as a reference #5720 and implement this change?

Sure. I will refer and make the changes to my PR.

Krishna-Prasad-P-V commented:
@majetideepak - I am getting error in `VariantSerializationTest.opaqueToString` because function `fromKindToScalerType` doesn't return the scalar type for OPAQUE kind. This is causing the below error.
```
E0720 02:47:38.046008 736355 Exceptions.h:68] Line: /root/dev/velox/velox/type/Type.cpp:828, Function:fromKindToScalerType, Expression:  Kind is not a scalar type: OPAQUE, Source: USER, ErrorCode: UNSUPPORTED
unknown file: Failure
C++ exception with description "Exception: VeloxUserError
Error Source: USER
Error Code: UNSUPPORTED
Reason: Kind is not a scalar type: OPAQUE
Retriable: False
Function: fromKindToScalerType
File: /root/dev/velox/velox/type/Type.cpp
Line: 828
Stack trace:
Stack trace has been disabled.Use --velox_exception_user_stacktrace=true to enable it.
" thrown in the test body.
[  FAILED  ] VariantSerializationTest.opaqueToString (1 ms)
[----------] 1 test from VariantSerializationTest (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] VariantSerializationTest.opaqueToString
```

Krishna-Prasad-P-V commented:
@majetideepak -I have doubts in removing `friend std::ostream& operator<<(std::ostream& stream, const variant& k)` from Variant.h.  Should we define it like this ? I see that `inferType()` can get us all scalar and non-scalar TypeKinds.
```
  friend std::ostream& operator<<(std::ostream& stream, const variant& k) {
    auto type = k.inferType();
    stream << k.toJson(type);
    return stream;
  }
``` 

Also, with reference to `pyToJson` declaration, should we use something like above. What happens when OPAQUE or UNKNOWN is used in this case? I think it will return error from `fromKindToScalerType` function. Let me know if this doubt of mine is valid here and if we can proceed with this.
```
  // Used by python binding, do not change signature.
  std::string pyToJson() const {
    const auto type = fromKindToScalerType(kind_);
    return toJson(type);
  }
```

majetideepak commented:
> I see that `inferType()` can get us all scalar and non-scalar TypeKinds.

We cannot get logical types with `inferType()`. So it would be incorrect to use that. That we we need to deprecate `toJson()` API.
Since `pyToJson()` code comment says we cannot modify the API, let's use `inferType()` for now. I will find out more about this.
For `operator<<(std::ostream& stream, const variant& k)`, we have to replace it with `toJson(TypePtr)`. I see that it gets used in QueryAssertions and I am pretty sure Logical Types like Decimal will not be displayed correctly there.

majetideepak commented on velox/type/Variant.h:
```diff
@@ -539,7 +540,8 @@ class variant {
   }
 
   friend std::ostream& operator<<(std::ostream& stream, const variant& k) {
```
We have to remove this operator as well as I explained in  https://github.com/facebookincubator/velox/pull/5406#issuecomment-1644486046

Krishna-Prasad-P-V replied:
Ok. I missed it. Let me remove and run the local tests.

Krishna-Prasad-P-V replied:
Fixed.

Krishna-Prasad-P-V replied:
Fixed.

majetideepak commented:
@Krishna-Prasad-P-V, using `inferType()` for an element of `MaterializedRow` (which is also a variant) is the same as using `toJson()` in `operator<<(std::ostream& stream, const variant& k)`. As mentioned here https://github.com/facebookincubator/velox/pull/5406#issuecomment-1644486046, we do not get logical types from the `inferType` call. We need to get the proper type from the context. For the MaterializedRow, you will get the proper type where the MaterializedRow is being created.

Krishna-Prasad-P-V commented:
> where the MaterializedRow is being created.

@majetideepak To test this with hugeint especially, I should be able to use this test in `QueryAssertionsTest.cpp`, right? Any suggestions? On running, currently with my old code change it will throw an error.
```
TEST_F(QueryAssertionsTest, nonNullDecimalValue) {
  auto data = makeRowVector({
      makeFlatVector<int128_t>(5, DECIMAL(20,2)),
  });
  createDuckDbTable({data});

  auto plan = PlanBuilder().values({data}).project({"c0"}).planNode();
  assertQuery(plan, "SELECT c0 FROM tmp");

  EXPECT_NONFATAL_FAILURE(
      assertQuery(plan, "SELECT c0 + 1 FROM tmp"),
      "1 extra rows, 1 missing rows");
}
```

majetideepak commented:
@Krishna-Prasad-P-V You are on the right path. The test below will show you the issue. You can probably fix this in a separate PR first.
```
TEST_F(QueryAssertionsTest, nonNullDecimalValue) {
  auto data = makeRowVector({
      makeFlatVector<int64_t>({10, 20}, DECIMAL(12,2)),
  });
  createDuckDbTable({data});

  auto plan = PlanBuilder().values({data}).project({"c0"}).planNode();
  assertQuery(plan, "SELECT c0 + 1 FROM tmp");
}
```
The output is shown in terms of the physical type int64_t, but not the logical decimal type.
```
Expected 2, got 2
2 extra rows, 2 missing rows
2 of extra rows:
	10
	20

2 of missing rows:
	110
	120
```

Krishna-Prasad-P-V commented:
> 

@majetideepak - So you recommend to fix this later. Then I think we could get this fixed by using `toJson(fromKindToScalerType(row[i].kind()))` within printRow method. Let me know if this is the right approach as you expected, instead of using `inferType()`.

majetideepak commented:
> Then I think we could get this fixed by using `toJson(fromKindToScalerType(row[i].kind()))` within printRow method

inferType() and above are equivalent. We need to fix the output in the comment below first. https://github.com/facebookincubator/velox/pull/5406#issuecomment-1655903054

You will have to plumb the TypePtr from the expected result vector.
The right output is
```
Expected 2, got 2
2 extra rows, 2 missing rows
2 of extra rows:
	0.10
	0.20

2 of missing rows:
	1.10
	1.20
```

Krishna-Prasad-P-V commented:
> vector

@majetideepak - So, I see that the result of a query `SELECT c0 FROM tmp` is obtained from method `DuckDbQueryRunner::execute` and returned as `materialize(dataChunk.get(), resultRowType)`. But, the doubt I have is that eve thought the table tmp is create as `c0 DECIMAL(12,2)`, we are not getting the result of the select query as DECIMAL(12,2). This was verified in this execute method and I found that `duckDbResult->Fetch()` is giving us the values 110 and 120, which is wrong if the result is stored correctly in the table. Is the record not saved correctly in table as decimal rather bigint was used ?

majetideepak commented:
> which is wrong if the result is stored correctly in the table. Is the record not saved correctly in table as decimal rather bigint was used ?

This relates to how Decimal values are stored. See: https://facebookincubator.github.io/velox/develop/types.html#logical-types

Krishna-Prasad-P-V commented:
@majetideepak - I was able to retrieve the precision and scale from expectedRows and create TypePtr from it. And, the output is now visible.
```
/root/dev/velox/velox/exec/tests/utils/QueryAssertions.cpp:1127: Failure
Failed
Expected 2, got 2
2 extra rows, 2 missing rows
2 of extra rows:
        0.10
        0.20

2 of missing rows:
        1.10
        1.20
```

majetideepak commented:
@Krishna-Prasad-P-V please open a separate PR to fix QueryAssertions as those changes are independent.
